### PR TITLE
fixed distribution of dependencies among tuning parameters

### PR DESCRIPTION
**Pull request by johanneslenfers**
## Fixed issue #194
Distribute constraints among tuning parameters creating an implicit dependency graph. 

## Notes to Solution
- removed special case of `RangeConstraint`
- implicit dependency graph is now created without cycles
- added function to check a distribution for cycles
- automatic config file generation works again

### CODE REVIEW

File: src/main/scala/rise/autotune/configFileGeneration.scala [hunk 1]
```diff
@@ -211,68 +211,34 @@ object configFileGeneration {
...
       Map[NatIdentifier, (Set[Constraint], Set[NatIdentifier])]()
-    parameters.foreach(param => {
+    parameters.toSeq.sortBy(_.name).foreach(param => {
       parametersWDC(param) = (Set.empty[Constraint], Set.empty[NatIdentifier])
     })
 
     // get parameters from constraint
     // check for given parameters in the given constraint
-    constraints.foreach(constraint => {
+    val test = constraints.toSeq.sortBy(_.toString)
```
Comment (Bastacyclop) [inline src/main/scala/rise/autotune/configFileGeneration.scala:220]: is this used?

File: src/test/scala/rise/autotuning.scala [hunk 2]
```diff
@@ -119,6 +119,37 @@ class autotuning extends test_util.Tests {
   }
   // scalastyle:on
 
+
+  // function to check cycle in parameter dependencies
+  def check_no_cycle(distribution: Map[NatIdentifier,
```
Comment (Bastacyclop) [inline src/test/scala/rise/autotuning.scala:124]: is this function only useful for testing, or should it be moved to `src/main`?

File: src/test/scala/rise/autotuning.scala [hunk 3]
```diff
@@ -119,6 +119,37 @@ class autotuning extends test_util.Tests {
...
+              })
+            }
+          }
+        }
+      }
+    }
+
+    distribution.exists(param => {
+      check_no_cycle_rec(param._1, param._2._2, distribution)
+    })
```
Comment (Bastacyclop) [inline src/test/scala/rise/autotuning.scala:150]: I am not sure to understand your algorithm here. There is no cycle if there exists a parameter which does not induce a cycle? Would this example pass the check:
```
A depends on B
B depends on A
C depends on D
```
I am also surprised that you do not use a set of visited nodes/parameters in your algorithm.

File: src/main/scala/rise/autotune/configFileGeneration.scala (hunk 1 above)
Comment (johanneslenfers) [reply src/main/scala/rise/autotune/configFileGeneration.scala]: removed

File: src/test/scala/rise/autotuning.scala (hunk 2 above)
Comment (johanneslenfers) [reply src/test/scala/rise/autotuning.scala]: I moved it to the config file generation of the auto tuning, where the parameter distribution is implemented.

File: src/test/scala/rise/autotuning.scala (hunk 3 above)
Comment (johanneslenfers) [reply src/test/scala/rise/autotuning.scala]: You are right. This should not work properly. I updated the algorithm.

Comment (johanneslenfers) [reply src/test/scala/rise/autotuning.scala]: The algorithm is updated.

File: src/main/scala/rise/autotune/configFileGeneration.scala [hunk 4]
```diff
@@ -299,4 +300,35 @@ object configFileGeneration {
...
+                           dependencies: Set[NatIdentifier],
+                           distribution: Map[NatIdentifier,
+                             (Set[constraints.Constraint], Set[NatIdentifier])],
+                           visited: Set[NatIdentifier]
+                          ): Boolean = {
+
+      dependencies.size match {
+        case 0 => true
+        case _ => dependencies.exists(dependency => {
+          dependency.name.equals(param.name) || visited.contains(dependency)
```
Comment (Bastacyclop) [inline src/main/scala/rise/autotune/configFileGeneration.scala:318]: You could avoid testing `dependency.name.equals(param.name)` by adding `param` to `visited` in the beginning.

File: src/main/scala/rise/autotune/configFileGeneration.scala [hunk 5]
```diff
@@ -299,4 +300,35 @@ object configFileGeneration {
...
+              distribution.apply(dependency)._2,
+              distribution,
+              visited + dependency)
+          })
+        }
+      }
+    }
+
+    distribution.forall(param =>
+      check_no_cycle_rec(param._1, param._2._2, distribution, Set.empty[NatIdentifier]))
```
Comment (Bastacyclop) [inline src/main/scala/rise/autotune/configFileGeneration.scala:332]: Complexity-wise, I think that what you are doing is roughly `O(V*(V + E))`: for every `V` vertices/parameters, you might visit the entire graph of `V` parameters and `E` edges/dependencies. If you care about performance (and also want to use a time-tested algorithm), you can use standard `O(V + E)` algorithms:
https://www.geeksforgeeks.org/detect-cycle-in-a-graph/
or
https://www.geeksforgeeks.org/detect-cycle-direct-graph-using-colors/

**Review by Bastacyclop — APPROVED** [review]

**Pull request merged**